backend/src/layer/logic/api_command/image_preprocessing/mod.rs

```diff
@@ -0,0 +1,105 @@
+//! Module for preprocessing uploaded images
+
+use image::{imageops::FilterType, io::Reader, DynamicImage, ImageError, ImageFormat};
+use std::io::Cursor;
+use thiserror::Error;
+
+/// Result returned on image preprocessing operations.
+pub type Result<T> = std::result::Result<T, ImagePreprocessingError>;
+
+/// Enum describing possible ways the image preprocessing can fail.
+#[derive(Error, Debug)]
+pub enum ImagePreprocessingError {
+    /// Error occurring when the image format could not be guessed out of the data and no mime type was provided.
+    #[error("Error guessing image format: {0}")]
+    FormatGuessError(std::io::Error),
+    /// Error occurring wile operating on the image.
+    #[error("Error during image operation: {0}")]
+    ImageError(#[from] ImageError),
+}
+
+/// Structure containing all information necessary for image preprocessing.
+#[derive(Debug, Clone, Copy)]
+pub struct ImagePreprocessingInfo {
+    /// Maximal width of images stored. Images get resized otherwise.
+    pub max_image_width: u32,
+    /// Maximal height of images stored. Images get resized otherwise.
+    pub max_image_height: u32,
+}
+
+/// Class for preprocessing uploaded images.
+#[derive(Debug)]
+pub struct ImagePreprocessor {
+    max_width: u32,
+    max_height: u32,
+}
+
+impl ImagePreprocessor {
+    /// Creates a new instance.
+    #[must_use]
+    pub const fn new(info: ImagePreprocessingInfo) -> Self {
+        Self {
+            max_width: info.max_image_width,
+            max_height: info.max_image_height,
+        }
+    }
+
+    /// Pre-process the given file to an image by reading and then down scaling it if to large.
+    /// # Panics
+    /// Should never panic.
+    /// # Errors
+    /// - Image type was not provided and could not be guessed
+    /// - Image could not be decoded from file
+    pub fn preprocess_image(
+        &self,
+        image_data: Vec<u8>,
+        image_type: Option<String>,
+    ) -> Result<DynamicImage> {
+        let max_width = self.max_width;
+        let max_height = self.max_height;
+
+        let mut reader = Reader::new(Cursor::new(image_data));
+        if let Some(format) = image_type.and_then(ImageFormat::from_mime_type) {
+            reader.set_format(format);
+        } else {
+            reader = reader
+                .with_guessed_format()
+                .map_err(ImagePreprocessingError::FormatGuessError)?;
+        }
+
+        // read image
+        let image = reader.decode()?;
+
```

Comment:
This resizing does not effect the image quality in terms of stretching?

Reply:
the aspect ration is preserved it that is what you mean